### PR TITLE
feat: EXC-1758: Evict sandboxes based on their RSS

### PR DESCRIPTION
**Pull request by berestovskyy**
This allows to increase the number of sandbox processes without risking the OOM as the total RSS size of sandboxes is still limited.

This PR also increases the number of sandbox processes to 5k.

### CODE REVIEW

File: rs/pocket_ic_server/src/pocket_ic.rs [hunk 1]
```diff
@@ -356,9 +357,10 @@ impl PocketIc {
             hypervisor_config.max_query_call_graph_instructions = instruction_limit;
         }
         // bound PocketIc resource consumption
-        hypervisor_config.embedders_config.min_sandbox_count = 0;
         hypervisor_config.embedders_config.max_sandbox_count = 64;
         hypervisor_config.embedders_config.max_sandbox_idle_time = Duration::from_secs(30);
+        hypervisor_config.embedders_config.max_sandboxes_rss_size =
+            NumBytes::new(2 * 1024 * 1024 * 1024);
```
Comment (adambratschikaye) [inline rs/pocket_ic_server/src/pocket_ic.rs:363]: I'm not sure if we need to impose a lower limit for pocket IC. If no one is complaining about it's memory usage the 64 count should still be fine, right?

**Review by adambratschikaye — APPROVED** [review]

**Review by mraszyk — APPROVED** [review]

File: rs/pocket_ic_server/src/pocket_ic.rs (hunk 1 above)
Comment (berestovskyy) [reply rs/pocket_ic_server/src/pocket_ic.rs:363]: Discussed offline, decided to go as is.

File: rs/canister_sandbox/src/replica_controller/sandboxed_execution_controller.rs [hunk 2]
```diff
@@ -1169,6 +1176,10 @@ impl SandboxedExecutionController {
                     metrics
                         .sandboxed_execution_subprocess_rss
                         .observe(process_rss as f64);
+                    if process_rss > 0 {
```
Comment (alexandru-uta) [inline rs/canister_sandbox/src/replica_controller/sandboxed_execution_controller.rs:1179]: The `process_rss` includes the `memfd_rss`, which is basically included in the page delta, which I think is already accounted for by the replica. I assume that we can get away without counting the `memfd_rss`, correct?

**Review by alexandru-uta — CHANGES_REQUESTED** [review]

File: rs/config/src/embedders.rs [hunk 3]
```diff
@@ -42,18 +42,18 @@ const DEFAULT_WASMTIME_RAYON_COMPILATION_THREADS: usize = 10;
...
-pub(crate) const DEFAULT_MAX_SANDBOX_COUNT: usize = 2_000;
+pub(crate) const DEFAULT_MAX_SANDBOX_COUNT: usize = 5_000;
 
 /// A sandbox process may be evicted after it has been idle for this
 /// duration and sandbox process eviction is activated.
 pub(crate) const DEFAULT_MAX_SANDBOX_IDLE_TIME: Duration = Duration::from_secs(30 * 60);
 
+/// Sandbox processes may be evicted if their total RSS memory size
+/// exceeds 100 GiB.
+pub(crate) const DEFAULT_MAX_SANDBOXES_RSS_SIZE: NumBytes = NumBytes::new(100 * 1024 * 1024 * 1024);
```
Comment (alexandru-uta) [inline rs/config/src/embedders.rs:55]: Are you sure we can afford 100 GB worth of sandboxes?

File: rs/canister_sandbox/src/replica_controller/sandbox_process_eviction.rs [hunk 4]
```diff
@@ -26,30 +28,29 @@ pub(crate) struct EvictionCandidate {
 /// 4. Return the evicted candidates.
 pub(crate) fn evict(
     mut candidates: Vec<EvictionCandidate>,
-    min_count_threshold: usize,
+    mut total_rss: NumBytes,
```
Comment (dsarlis) [inline rs/canister_sandbox/src/replica_controller/sandbox_process_eviction.rs:31]: Very minor point: would it make more sense to not accept a mutable `total_rss` here but rather that the function maintains a "running" total rss while it's evicting? It seems somewhat cleaner to me but I'm also ok with this.

**Review by dsarlis — APPROVED** [review]

File: rs/config/src/embedders.rs (hunk 3 above)
Comment (berestovskyy) [reply rs/config/src/embedders.rs:55]: That's the current state of the mainnet (2k sandboxes times 50 MiB each). Metrics indicate that nodes have a minimum of 270+ GiB of available memory, with up to 2k of active sandboxes.

As discussed offline, we should prioritize work to mitigate risks.

Comment (alexandru-uta) [reply rs/config/src/embedders.rs]: I don't think this is the way it works; we had that document with max memory usage per component for a reason; we're all of a sudden doubling something without taking anything back.

**Review by alexandru-uta — APPROVED** [review]
thanks.

File: rs/config/src/embedders.rs (hunk 3 above)
Comment (berestovskyy) [reply rs/config/src/embedders.rs]: Discussed offline. The document is outdated, as two weeks ago we've doubled the number of sandboxes to 2k on mainnet (up to 50 MiB each totaling 100 GiB). Anyway, reduced the limit to 50 GiB.

**Pull request merged**